metal/label_model/class_balance.py

```diff
@@ -0,0 +1,136 @@
+from itertools import product
+
+import numpy as np
+import torch
+from torch import nn, optim
+
+
+class ClassBalanceModel(nn.Module):
+    """A model for learning the class balance, P(Y=y), given a subset of LFs
+    which are *conditionally independent*, i.e. \lambda_i \perp \lambda_j | Y,
+    for  i != j.
+
+    Learns the model using a tensor factorization approach.
+
+    Note: This approach can also be used for estimation of the LabelModel, may
+    want to later refactor and expand this class.
+    """
+
+    def __init__(self, k, abstains=True, config=None):
+        super().__init__()
+        self.config = config
+        self.k = k
+        self.abstains = abstains
+        self.k_0 = 0 if self.abstains else 1
+        self.k_lf = k + 1 if self.abstains else k
+
+        # Estimated quantities (np.array)
+        self.cond_probs = None
+        self.class_balance = None
+
+    def _get_overlaps_tensor(self, L):
+        """Transforms the input label matrix to a three-way overlaps tensor.
+
+        Args:
+            L: (np.array) An n x m array of LF output labels, in {0,...,k} if
+                self.abstains, else in {1,...,k}, generated by m conditionally
+                independent LFs on n data points
+
+        Outputs:
+            O: (torch.Tensor) A (m, m, m, k, k, k) tensor of the label-specific
+            overlap rates.
+        """
+        n, m = L.shape
+
+        # Convert from a (n,m) matrix of ints to a (k_lf, n, m) indicator tensor
+        LY = np.array(
+            [np.where(L == y, 1, 0) for y in range(self.k_0, self.k + 1)]
+        )
+
+        # Form the three-way overlaps matrix
+        O = np.einsum("abc,dbe,fbg->cegadf", LY, LY, LY) / n
+        return torch.from_numpy(O).float()
+
+    def get_mask(self, m):
+        """Get the mask for the three-way overlaps matrix O, which is 0 when
+        indices i,j,k are not unique"""
+        mask = torch.ones((m, m, m, self.k_lf, self.k_lf, self.k_lf)).byte()
+        for i, j, k in product(range(m), repeat=3):
+            if len(set((i, j, k))) < 3:
+                mask[i, j, k, :, :, :] = 0
+        return mask
+
+    @staticmethod
+    def get_loss(O, Q, mask):
+        # Main constraint: match empirical three-way overlaps matrix
+        # (entries O_{ijk} for i != j != k)
+        diffs = (O - torch.einsum("aby,cdy,efy->acebdf", [Q, Q, Q]))[mask]
+        return torch.norm(diffs) ** 2
+
+    def train_model(self, L=None, O=None, lr=1, max_iter=1000, verbose=False):
+
+        # Get overlaps tensor if L provided else use O directly (e.g. for tests)
+        if O is not None:
+            pass
+        elif L is not None:
+            O = self._get_overlaps_tensor(L)
+        else:
+            raise ValueError("L or O required as input.")
+        self.m = O.shape[0]
+
+        # Compute mask
+        self.mask = self.get_mask(self.m)
+
+        # Initialize parameters
+        self.Q = nn.Parameter(
```

Comment:
That is a whole lot of converting and wrapping. Does torch.random not work here?

Reply:
Fixed!